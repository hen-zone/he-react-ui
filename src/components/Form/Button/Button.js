// @flow

/**
 *
 * Button
 *
 */

import classnames from 'classnames';
import React, { Fragment } from 'react';
import { Link } from 'react-router-dom';
import Icon from '../../Icon';
import styles from './Button.scss';
import ButtonSpinner from './ButtonSpinner';

function InnerButton(props: {
  to: ?string,
  withRouter: boolean,
  submit: boolean,
  children: any,
  className: string,
  onClick: ?Function,
}) {
  const { to, withRouter, submit, children, className, onClick } = props;
  if (!to) {
    return (
      <button
        className={className}
        onClick={onClick}
        type={submit ? 'submit' : 'button'}
      >
        {children}
      </button>
    );
  }

  return withRouter ? (
    <Link to={to} className={className}>
      {children}
    </Link>
  ) : (
    <a href={to} className={className}>
      {children}
    </a>
  );
}

type Props = {
  submit: boolean,
  color?: 'teal' | 'blue' | 'green' | 'red' | 'white',
  link?: boolean,
  small?: boolean,
  large?: boolean,
  onClick: Function,
  keyline?: boolean,
  children?: string | Array<any>,
  disabled?: boolean,
  icon?: string,
  iconLeft?: any,
  iconRight?: any,
  squared?: boolean,
  style?: any,
  submitting?: boolean,
  done?: boolean,
  className?: string,
  to?: string,
  withRouter: boolean,
};

class Button extends React.Component<Props> {
  static defaultProps = {
    submit: false,
    link: false,
    color: 'teal',
    icon: '',
    keyline: false,
    disabled: false,
    squared: false,
    onClick: () => null,
    withRouter: false,
  };

  handleClick = (event: SyntheticEvent<*>) => {
    if (!this.props.disabled) {
      this.props.onClick(event);
    }
  };

  render() {
    const {
      children,
      className,
      color,
      disabled,
      icon,
      iconLeft,
      iconRight,
      keyline,
      link,
      small,
      large,
      squared,
      style,
      submit,
      submitting,
      done,
      to,
      withRouter,
    } = this.props;

    const buttonClasses = classnames(
      styles.button,
      color && !done && styles[color],
      {
        [styles.disabled]: disabled,
        [styles.keyline]: keyline,
        [styles.link]: link,
        [styles.squared]: squared,
        [styles.submitting]: submitting,
        [styles.done]: done,
      },
      className,
    );

    const containerClasses = classnames(styles.buttonContainer, {
      [styles.small]: small,
      [styles.large]: large,
    });

    const submittingIcon = submitting ? (
      <div className={styles.iconCenter}>
        <ButtonSpinner />
      </div>
    ) : null;

    const statusIcon = done ? (
      <div className={styles.iconCenter}>
        <Icon name="Tick" />
      </div>
    ) : (
      submittingIcon
    );

    return (
      <div className={containerClasses} style={style}>
        <InnerButton
          className={buttonClasses}
          submit={submit}
          to={to}
          withRouter={withRouter}
          onClick={this.handleClick}
        >
          {statusIcon || (
            <Fragment>
              {iconLeft && <div className={styles.iconLeft}>{iconLeft}</div>}

              <div className={styles.content}>
                {children}

                {icon && <Icon className={styles.legacyIcon} name={icon} />}
              </div>

              {iconRight && <div className={styles.iconRight}>{iconRight}</div>}
            </Fragment>
          )}
        </InnerButton>
      </div>
    );
  }
}

export default Button;
