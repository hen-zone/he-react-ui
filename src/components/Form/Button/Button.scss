@import '../../../styles/base';

.buttonContainer {
  display: inline-flex;
  height: 48px;
  min-width: 180px;

  align-items: center;
  justify-content: center;

  &+.buttonContainer {
    margin-left: 24px;
  }

  @include breakpoint(small) {
    min-width: 0;
    width: 100%;
    &+.buttonContainer {
      margin-left: 0;
      margin-top: 16px;
    }
  }

  &.small {
    height: 30px;
    min-width: 100px;
    @include breakpoint(small) {
      min-width: 0;
    }
  }

  &.large {
    min-width: 280px;
    @include breakpoint(small) {
      min-width: 0;
    }
  }
}

.button {
  display: flex;
  flex-grow: 1;
  height: 100%;
  flex-basis: 48px;
  white-space: nowrap;
  cursor: pointer;
  color: $white;
  padding: 0 36px;
  border: none;
  border-radius: 48px;
  text-align: left;
  position: relative;
  font-size: $input-size;
  font-family: $font-stack;
  outline: none;
  font-weight: $semi-bold;
  text-decoration: none;
  user-select: none;
  align-items: center;
  justify-content: center;
  width: 100%;

  transition-property: flex-grow;
  transition-duration: 0.3s;

  &.submitting {
    flex-grow: 0.01; // working around a chrome bug
    padding: 0;
    pointer-events: none;
  }

  &.done {
    background-color: $green;
    pointer-events: none;
  }

  &:not(.disabled):not(.submitting):not(.done):hover {
    filter: brightness(92%);
  }
  &:not(.disabled):not(.submitting):not(.done):active {
    filter: brightness(84%);
  }
}

.content {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}


.teal {
  background: $teal;
  &:active, &:hover, &:focus {
    background: $teal;
  }
}

.blue {
  background: $blue;
  &:active, &:hover, &:focus {
    background: $blue;
  }
}

.green {
  background: $green;
  &:active, &:hover, &:focus {
    background: $green;
  }
}

.red {
  background: $red;
  &:active, &:hover, &:focus {
    background: $red;
  }
}

.white {
  background: $white;
  color: $teal_dark_20;
  &:active, &:hover, &:focus {
    background: $white;
  }
}

.link {
  background: none;
  border: none;
  padding: 0;
  height: auto;
  &:active, &:hover, &:focus {
    background: none;
  }
}

.link, .keyline:not(:hover):not(:active) {
  border-width: 2px;

  &.teal {
    color: $teal;
    border-color: $teal;
  }

  &.blue {
    color: $blue;
    border-color: $blue;
  }

  &.green {
    color: $green;
    border-color: $green;
  }

  &.red {
    color: $red;
    border-color: $red;
  }

  &.white {
    color: $white;
    border-color: $white;
  }
}

.squared {
  border-radius: 2px;
}

.iconContainer {
  position: absolute;
  height: 24px;
  width: 24px;
  margin-top: -12px;
  top: 50%;

  >* {
    width: 100%;
    height: 100%;
  }
}

.iconLeft {
  @extend .iconContainer;
  left: 12px;
}

.iconRight {
  @extend .iconContainer;
  right: 12px;
}

.iconCenter {
  @extend .iconContainer;
  left: 50%;
  margin-left: -12px;
}

.legacyIcon {
  fill: currentColor;
  height: 24px;
  width: 24px;
  position: absolute;
  right: 12px;
  top: 12px;
}

.keyline {
  &:not(:hover):not(:active) {
    border: solid 1px;
    background: none;
  }
}

.disabled {
  cursor: not-allowed;
  &:not(.link) {
    background: $grey_4 !important;
    &:active, &:hover, &:focus {
      background: $grey_4 !important;
    }
  }
  color: $white !important;
  border: none !important;
}
