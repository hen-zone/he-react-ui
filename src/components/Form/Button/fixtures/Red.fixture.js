import Button from '../'

export default {
  component: Button,
  props: {
    color: 'red'
  },
  children: 'Button',
  simulateSubmission: true
}
