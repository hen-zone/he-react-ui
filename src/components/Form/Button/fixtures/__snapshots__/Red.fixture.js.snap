// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`anonymous 1`] = `
<Button
  color="red"
  disabled={false}
  done={false}
  icon=""
  keyline={false}
  link={false}
  newWindow={false}
  onClick={[Function]}
  squared={false}
  submit={false}
  submitting={false}
  withRouter={false}
>
  <div
    className="buttonContainer"
  >
    <InnerButton
      className="button red"
      newWindow={false}
      onClick={[Function]}
      submit={false}
      withRouter={false}
    >
      <button
        className="button red"
        onClick={[Function]}
        type="button"
      >
        <div
          className="content"
        >
          Button
        </div>
      </button>
    </InnerButton>
  </div>
</Button>
`;
